File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/sizing/SizeConfig.swift
//
//  SizeConfig.swift
//  StyleEngine/sizing — Phase 3.
//
//  Canonical per-component sizing config. Each axis carries a full
//  `LengthValue` (not a CGFloat) so unit-resolution can defer until
//  the applier knows the parent width. Logical sides (BlockSize /
//  InlineSize etc.) are resolved to physical at extract time assuming
//  LTR horizontal writing mode — the only mode iOS targets today.
//
//  AspectRatio lives on its own field because its IR shape is disjoint
//  from LengthValue (see AspectRatioValue.swift).
//

// Foundation only — no SwiftUI dependency so SizeConfig stays testable.
import Foundation

// css-sizing-3 §3 `box-sizing` keyword. Only the two spec values exist —
// the TRI-STATE the sizing lane needs (unset ≠ content-box!) is carried
// by Optional wrapping on SizeConfig.boxSizing below: `nil` means "the IR
// never declared box-sizing", which must keep today's border-box frames
// byte-stable (the whole width+padding fixture corpus is captured against
// the web harness's `* { box-sizing: border-box }` reset).
enum BoxSizingKeyword: Equatable {
    // Declared size = content box; frame = content + padding + border.
    case contentBox
    // Declared size = border box (the pre-existing iOS chain behaviour).
    case borderBox
}

// Unified sizing bundle. A `nil` LengthValue means the property was not
// present in the IR; the applier can use that to skip attaching a
// `.frame(...)` modifier on that axis. `.none` / `.auto` / etc. are
// distinct states carried inside the LengthValue itself.
struct SizeConfig: Equatable {
    // Physical width / height.
    var width: LengthValue? = nil
    var height: LengthValue? = nil

    // Min/Max constraints. `.none` arrives here when `max-*: none`.
    var minWidth: LengthValue? = nil
    var maxWidth: LengthValue? = nil
    var minHeight: LengthValue? = nil
    var maxHeight: LengthValue? = nil

    // `aspect-ratio` — own type because IR shape diverges from lengths.
    // Nil means the property was absent (distinct from `.isAuto`).
    var aspectRatio: AspectRatioValue? = nil

    // css-values-5 calc-size() typed values (wave 42 lane W3) — the two
    // slots iOS consumes (preferred width/height). A slot carries EITHER
    // its LengthValue OR its calc value, never both (SizeExtractor routes
    // exclusively). Min*/Max* calc-size deliberately has NO slot here —
    // see SizeExtractor's calc-size comment for the measured record (the
    // flex family PASSES on the intrinsic fallback those slots keep).
    var widthCalc: CalcSizeValue? = nil
    var heightCalc: CalcSizeValue? = nil

    // css-sizing-3 §3 `box-sizing`. Nil = the IR never declared it (the
    // border-box status quo must not change); `.contentBox` makes the
    // applier inflate explicit width/height by the padding + border
    // bands so the FRAME equals content + padding + border.
    var boxSizing: BoxSizingKeyword? = nil

    // True when any sizing field is populated. Callers use this to skip
    // attaching SizeApplier altogether when the bag is empty.
    // `boxSizing` is deliberately EXCLUDED: box-sizing only changes how
    // definite width/height resolve (css-sizing-3 §3) — with no size to
    // reinterpret the applier has nothing to do, so a lone box-sizing
    // declaration must not force the modifier onto the chain.
    var hasAny: Bool {
        width != nil || height != nil ||
        minWidth != nil || maxWidth != nil ||
        minHeight != nil || maxHeight != nil ||
        aspectRatio != nil ||
        // calc-size slots size the box too — without these the applier's
        // fast path would skip a component whose ONLY sizing is a typed
        // calc-size width (exactly the calc-size-min-max inner-div shape).
        widthCalc != nil || heightCalc != nil
    }
}

File: testing/web/src/style/engine/sizing/SizeConfig.ts
// SizeConfig.ts — typed record holding every sizing value a component sets.
// One shared config carries both the physical CSS axes (width/height/min-*/max-*)
// and the logical counterparts (block-size / inline-size / min-* / max-*),
// because modern browsers accept the logical longhands directly and the CSS
// engine resolves them against writing-mode/direction at layout time.
//
// AspectRatio lives here too so a single extract/apply pair covers every
// sizing property the Phase-3 migration owns.

// Reuse the Phase-1 discriminated-union so we speak one length alphabet across
// the entire engine (exact px, relative unit, auto, none, intrinsic, calc, ...).
import type { LengthValue } from '../core/types/LengthValue';
// AspectRatio has its own (ratio, isAuto) shape — kept separate to avoid
// polluting LengthValue.
import type { AspectRatioValue } from './AspectRatioValue';

// Every field is optional — the extractor only populates keys whose IR
// property actually appeared in the component's properties array.
export interface SizeConfig {
  // Physical sizing — width / height longhands from IR.
  width?: LengthValue;
  height?: LengthValue;
  minWidth?: LengthValue;
  maxWidth?: LengthValue;
  minHeight?: LengthValue;
  maxHeight?: LengthValue;
  // Logical sizing — block-size / inline-size longhands.  These map to the
  // `blockSize` / `inlineSize` CSS camelCase keys directly; CSS resolves
  // against the current writing-mode.
  blockSize?: LengthValue;
  inlineSize?: LengthValue;
  minBlockSize?: LengthValue;
  maxBlockSize?: LengthValue;
  minInlineSize?: LengthValue;
  maxInlineSize?: LengthValue;
  // Aspect ratio — `null` / missing means "don't emit".  An `isAuto:true`
  // value round-trips to the CSS keyword `auto`.
  aspectRatio?: AspectRatioValue;
}

// IR property-type strings the sizing pipeline owns.  Used both by the
// renderer (to gate legacy vs engine path) and by the extractor's switch.
export const SIZE_PROPERTY_TYPES = [
  // Physical sizing (from irmodels/properties/spacing/).
  'Width', 'Height',
  'MinWidth', 'MaxWidth', 'MinHeight', 'MaxHeight',
  // Logical sizing (from irmodels/properties/sizing/).
  'BlockSize', 'InlineSize',
  'MinBlockSize', 'MaxBlockSize', 'MinInlineSize', 'MaxInlineSize',
  // AspectRatio — lives alongside sizing; shape handled by AspectRatioValue.
  'AspectRatio',
] as const;

// Tuple literal → union type — lets the extractor's switch be exhaustive.
export type SizePropertyType = (typeof SIZE_PROPERTY_TYPES)[number];
